### PR TITLE
SDI-44 get visit endpoint

### PR DESCRIPTION
**Pull request by lbennett-moj**
removed optionals to match db non nullable columns

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/VisitResponse.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package uk.gov.justice.digital.hmpps.nomisprisonerapi.data
+
+import com.fasterxml.jackson.annotation.JsonInclude
+import io.swagger.v3.oas.annotations.media.Schema
+import java.time.LocalDateTime
+import javax.validation.constraints.NotEmpty
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@Schema(description = "Visit information")
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/VisitResponse.kt:9]: first guess at migration requirements, will revisit as necessary

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/VisitResourceIntTest.kt [hunk 2]
```diff
@@ -76,15 +78,19 @@ class VisitResourceIntTest : IntegrationTestBase() {
...
+    @BeforeEach
+    internal fun setUpData() {
+      createPrisoner()
+    }
+
+    @AfterEach
+    internal fun deleteData() {
+      repository.delete(offender)
+      repository.delete(threePeople)
+    }
```
Comment (lbennett-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/VisitResourceIntTest.kt:94]: moving data requirements into nested classes to allow specific test data creation

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/VisitResourceIntTest.kt [hunk 3]
```diff
@@ -349,6 +366,91 @@ class VisitResourceIntTest : IntegrationTestBase() {
...
+      val visitId = offender.bookings[0].visits[0].id
+      val visit = webTestClient.get().uri("/visits/$visitId")
+        .headers(setAuthorisation(roles = listOf("ROLE_READ_NOMIS")))
+        .exchange()
+        .expectStatus().isOk
+        .expectBody(VisitResponse::class.java)
+        .returnResult().responseBody!!
+
+      assertThat(visit.visitStatus.code).isEqualTo("SCH")
+      assertThat(visit.visitType.code).isEqualTo("SCON")
```
Comment (lbennett-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/VisitResourceIntTest.kt:410]: will look at testing the json response in a future commit

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt [hunk 4]
```diff
@@ -93,13 +95,13 @@ class VisitService(
       .orElseThrow(BadDataException("Invalid cancellation reason: ${visitDto.outcome}"))
 
     val visitOrder = visit.visitOrder
-    if (visit.visitStatus?.code == "CANC") {
+    if (visit.visitStatus.code == "CANC") {
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt:98]: Yay this'll be much neater

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt [hunk 5]
```diff
@@ -263,6 +265,24 @@ class VisitService(
...
+        endDateTime = LocalDateTime.from(this.endDateTime),
+        visitType = VisitResponse.CodeDescription(this.visitType.code, this.visitType.description),
+        visitStatus = VisitResponse.CodeDescription(this.visitStatus.code, this.visitStatus.description),
+        agencyInternalLocation = this.agencyInternalLocation?.let { VisitResponse.CodeDescription(it.locationCode!!, it.description!!) },
+        commentText = this.commentText,
+        visitorConcernText = this.visitorConcernText,
+        visitors = this.visitors.filter { visitor -> visitor.person != null }.map { visitor -> VisitResponse.Visitor(visitor.person!!.id, visitor.groupLeader) }
+      )
+    } ?: throw NotFoundException("visit id $visitId")
+  }
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt:285]: Are you deliberately keeping the optional "this." ?

Comment (lbennett-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt:285]: yes - might be easier to discuss

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt [hunk 6]
```diff
@@ -225,8 +227,8 @@ class VisitService(
     return Visit(
       offenderBooking = offenderBooking,
       visitDate = LocalDate.from(visitDto.startDateTime),
-      startTime = visitDto.startDateTime,
-      endTime = LocalDateTime.of(LocalDate.from(visitDto.startDateTime), visitDto.endTime),
+      startDateTime = visitDto.startDateTime,
+      endDateTime = LocalDateTime.of(LocalDate.from(visitDto.startDateTime), visitDto.endTime),
       visitType = visitType,
       visitStatus = visitStatusRepository.findById(VisitStatus.pk("SCH")).orElseThrow(),
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt:233]: curious, not sure I understand this visitStatus, isn't it always returning scheduled?

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt (hunk 6 above)
Comment (lbennett-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt:233]: yeah - doesn't seem right.  I would have thought using the value passed by the sync service (albeit always "SCH" would be correct

**Pull request merged**